src/eval/lazy.rs

```diff
@@ -35,8 +35,71 @@ pub struct ThunkData {
 /// - A revertible thunk, that can be restored to its original expression. Used to implement
 ///   recursive merging of records and overriding (see the
 ///   [RFC overriding](https://github.com/tweag/nickel/pull/330)). A revertible thunks optionally
-///   stores the set of recusive fields it depends on. See the [`crate::transform::free_vars`] for more
-///   details.
+///   stores the set of recusive fields it depends on (see [`crate::transform::free_vars`]).
+///
+/// # Revertible thunks
+///
+/// Recursive records are better understood as being functions. For example, the recursive record:
+///
+/// ```nickel
+/// {
+///   foo = bar + baz + 1,
+///   bar = 2,
+///   baz = 1,
+/// }
+/// ```
+///
+/// is really a builtin representation of `fun self => {foo = self.bar + self.baz + 1, bar = 2, baz
+/// = 1}`. The interpreter computes a fixpoint when evaluating a `RecRecord` away to a `Record`,
+/// equivalent to `let rec fixpoint = repr fixpoint in fixpoint`. We use a different vocabulary and
+/// representations, but this is what we essentially do. We're just hiding the details from the
+/// user and pretend recursive records are data that can be used as a normal record, because it's
+/// both simple and natural (for the user).
+///
+/// A naive representation of recursive records could be to always keep recursive records in their
+/// functional representation, and re-compute the fixpoint at each record operation, such as field
+/// access. While correct, this would be wasteful, because the result of the fixpoint computations
+/// is actually the same throughout most record operations (put differently, `self` doesn't
+/// change). But we can't just store the computed `fixpoint` either in a normal thunk and forget
+/// about the original function, because upon merge, and more specifically upon recursive
+/// overriding happen, we do need to compute a new fixpoint.
+///
+/// Revertible thunks are a simple **memoization device** for the `fun self => ...` functional
+/// representation of a recursive record. First, note that there are several different but
+/// equivalent ways of encoding the example above as a function:
```

Comment:
As a rule of thumb I always recommend to not use "simple" or synonyms in any form of documentation :sweat_smile: 

At best it is unnecessary, at worst it can make the reader feel dumb.

Reply:
Hmm, I meant simple in the fact that it can only memoize on function call, on one argument, basically. As opposed to say memoizing function on `n` different entries. But I agree `simple` is not very specific here.